content/posts/2025/building-a-minimal-viable-armv7-emulator.md

```diff
@@ -315,18 +475,1399 @@ macro_rules! le32 {
+```
+
 # Dumping ELF segments into memory
 
-# Decoding ARMv7
+Since the only reason for parsing the elf headers is to know where to put what
+segment with which permissions, I want to quickly interject on why we have to
+put said segments at these specific addresses. The main reason is that all
+pointers, all offsets and pc related decoding has to be done relative to
+`Elf32_Ehdr.entry`. The linker also generated all instruction arguments
+according to this value.
+
+Before mapping each segment at its `Pheader::vaddr`, we have to understand:
+One doesn't simply `mmap` with `MAP_FIXED` or `MAP_NOREPLACE` into the virtual
+address `0x8000`. The Linux kernel won't let us, and rightfully so, `man mmap`
+says:
+
+> If addr is not NULL, then the kernel takes it as a hint about where to place
+> the mapping; on Linux, the kernel will pick a nearby page boundary (but
+> always above or equal to the value specified by /proc/sys/vm/mmap_min_addr) and
+> attempt to create the mapping there.
+
+And `/proc/sys/vm/mmap_min_addr` on my system is `u16::MAX` (2^16)-1=65535. So
+mapping our segment to `0x8000` (32768) is not allowed:
+
+```rust
+let segment = sys::mmap::mmap(
+    // this is only UB if dereferenced, its just a hint, so its safe here
+    Some(unsafe { std::ptr::NonNull::new_unchecked(0x8000 as *mut u8) }),
+    4096,
+    sys::mmap::MmapProt::WRITE,
+    sys::mmap::MmapFlags::ANONYMOUS
+        | sys::mmap::MmapFlags::PRIVATE
+        | sys::mmap::MmapFlags::NOREPLACE,
+    -1,
+    0,
+)
+.unwrap();
+```
+
+Running the above with our `vaddr` of `0x8000` results in:
+
+```text
+thread 'main' panicked at src/main.rs:33:6:
+called `Result::unwrap()` on an `Err` value: "mmap failed (errno 1): Operation not permitted
+(os error 1)"
+```
+
+It only works in elevated permission mode, which is something I dont want to
+run my emulator in.
+
+## Translating guest memory access to host memory access
+
+The obvious fix is to not mmap below `u16::MAX` and let the kernel choose where
+we dump our segment:
+
+```rust
+let segment = sys::mmap::mmap(
+    None,
+    4096,
+    MmapProt::WRITE,
+    MmapFlags::ANONYMOUS | MmapFlags::PRIVATE,
+    -1,
+    0,
+).unwrap();
+```
+
+But this means the segment of the process to emulate is not at `0x8000`, but
+anywhere the kernel allows. So we need to add a translation layer between guest
+and host memory: (If you're familiar with how virtual memory works, its similar
+but one more indirection)
+
+```text
++--guest--+
+| 0x80000 | ------------+
++---------+             |
+                        |
+                    Mem::translate
+                        |
++------host------+      |
+| 0x7f5b4b8f8000 | <----+
++----------------+
+```
+
+Putting this into rust:
+
+- `map_region` registers a region of memory and allows `Mem`
+  to take ownership for calling munmap on these segments
+  once it goes out of scope
+- `translate` does what the diagram above shows
```

Comment:
Würde es erklären zur Sicherheit wenn Menschen Diagramme nicht verstehen/sehen können: translate takes a Memory address call from the arm binary and translates it to the host memory address.